server/src/main/java/io/crate/protocols/postgres/PostgresWireProtocol.java

```diff
@@ -801,19 +804,30 @@ void handleSimpleQuery(ByteBuf buffer, final DelayableWriteChannel channel) {
             sendReadyForQuery(channel, TransactionState.IDLE);
             return;
         }
+        timeoutToken.failIfExpired(queryString, "parsing", System.currentTimeMillis() - startTime);
         CompletableFuture<?> composedFuture = CompletableFuture.completedFuture(null);
         for (var statement : statements) {
-            composedFuture = composedFuture.thenCompose(result -> handleSingleQuery(statement, queryString, channel));
+            composedFuture = composedFuture.thenCompose(result -> handleSingleQuery(statement, queryString, channel, timeoutToken));
         }
         composedFuture.whenComplete(new ReadyForQueryCallback(channel, TransactionState.IDLE));
```

Comment:
We currently rely on an exception for the timeout, but the `CompletableFuture` could also take a timeout directly using ` orTimeout`.

Reply:
Not sure we should have 2 places to deal with the timeout - current mechanism is protocol agnostic and everything happens in Session. 

Also, we need to handle happy path (no timeout) but still subtract parsing/planning time in the `Session.addStatementTimeout` to let exectue run less - so we anyway have to use session-based logic